piq/iw_ssim.py

```diff
@@ -0,0 +1,458 @@
+    down_pad = kernel.size(-1) - 1 - up_pad  # 5 -> 2, 4 -> 2
+    hi_x = lo_x
+
+    if x.size(-1) > 1:
+        upsampling_kernel = torch.tensor([[[[1., 0.]]]], dtype=x.dtype, device=x.device)
+        hi_x = F.conv_transpose2d(input=hi_x, weight=upsampling_kernel, stride=(1, 2), padding=0)
+        hi_x = F.pad(hi_x, pad=[up_pad, down_pad, 0, 0], mode='reflect')
+        hi_x = F.conv2d(input=hi_x, weight=kernel.unsqueeze(0), padding=0)[:, :, :, :x.size(-1)]
+    if x.size(-2) > 1:
+        upsampling_kernel = torch.tensor([[[[1.], [0.]]]], dtype=x.dtype, device=x.device)
+        hi_x = F.conv_transpose2d(input=hi_x, weight=upsampling_kernel, stride=(2, 1), padding=0)
+        hi_x = F.pad(hi_x, pad=[0, 0, up_pad, down_pad], mode='reflect')
+        hi_x = F.conv2d(input=hi_x, weight=kernel_t.unsqueeze(0), padding=0)[:, :, :x.size(-2), :]
+
+    hi_x = x - hi_x
+    return lo_x, hi_x
+
+
+def _ssim_per_channel(x: torch.Tensor, y: torch.Tensor, kernel: torch.Tensor,
+                      data_range: Union[float, int] = 1., k1: float = 0.01,
+                      k2: float = 0.03) -> Tuple[torch.Tensor, torch.Tensor]:
+    r"""Computes Structural Similarity (SSIM) index for X and Y per channel.
+
+    Args:
+        x: An input tensor. Shape :math:`(N, C, H, W)`.
+        y: A target tensor. Shape :math:`(N, C, H, W)`.
+        kernel: 2D Gaussian kernel.
+        data_range: Maximum value range of images (usually 1.0 or 255).
+        k1: Algorithm parameter, K1 (small constant, see [1]).
+        k2: Algorithm parameter, K2 (small constant, see [1]).
+            Try a larger K2 constant (e.g. 0.4) if you get a negative or NaN results.
+
+    Returns:
+        Tuple with Structural Similarity maps and Contrast maps.
+    """
+    if x.size(-1) < kernel.size(-1) or x.size(-2) < kernel.size(-2):
+        raise ValueError(f'Kernel size can\'t be greater than actual input size. Input size: {x.size()}. '
+                         f'Kernel size: {kernel.size()}')
+
+    c1 = (k1 * data_range) ** 2
+    c2 = (k2 * data_range) ** 2
+    n_channels = x.size(1)
+    mu_x = F.conv2d(x, weight=kernel, stride=1, padding=0, groups=n_channels)
+    mu_y = F.conv2d(y, weight=kernel, stride=1, padding=0, groups=n_channels)
+
+    mu_xx = mu_x ** 2
+    mu_yy = mu_y ** 2
+    mu_xy = mu_x * mu_y
+
+    sigma_xx = F.conv2d(x ** 2, weight=kernel, stride=1, padding=0, groups=n_channels) - mu_xx
+    sigma_yy = F.conv2d(y ** 2, weight=kernel, stride=1, padding=0, groups=n_channels) - mu_yy
+    sigma_xy = F.conv2d(x * y, weight=kernel, stride=1, padding=0, groups=n_channels) - mu_xy
+
+    sigma_xx = F.relu(sigma_xx)
+    sigma_yy = F.relu(sigma_yy)
+
+    # Contrast sensitivity (CS) with alpha = beta = gamma = 1.
+    cs_map = (2. * sigma_xy + c2) / (sigma_xx + sigma_yy + c2)
+
+    # Structural similarity (SSIM)
+    ss_map = (2. * mu_xy + c1) / (mu_xx + mu_yy + c1) * cs_map
+
+    return ss_map, cs_map
```

Comment:
What's the difference between this function and one defined in `ssim.py`? Can we reuse it?